Replace build-packages:watch to only rebuild changed package

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

The `build-packages:watch` npm script was added in #37802 to allow rebuilding the `@automattic/composite-checkout` package when it changed in the monorepo (and to also trigger a rebuild of calypso so that the modifications would be used). However, it causes a rebuild of _all_ packages any time a file inside a single package has changed. This can be a very slow process.

In this PR we introduce a new script that allows rebuilding only the packages which contained the changed file. This should be considerably faster. (In my tests it went from about 120 seconds down to about 10 seconds depending on the package changed).

This also ignores changes to Markdown files and files in test directories in `packages/`.

#### Testing instructions

- Run `yarn start` and wait for calypso to be started.
- Modify a source file inside the `packages` directory.
- Verify that you see a `rebuidling` message for the package which contains the changed file. 
- Verify that the change you made took effect by loading http://calypso.localhost:3000 and looking for your change. A good change to make is a `console.log()` statement for which it is easy to see the results.

sirbrillig committed:
Replace build-packages:watch to only rebuild changed package

sirbrillig commented:
One thing I'd like to figure out in addition to this change is if there's a way to prevent rebuilding calypso twice when a package change happens. One build starts when `yarn run prepare` occurs, and then another when the package rebuild completes. I'm not sure yet what triggers that first build.

scinos commented on bin/build-package-containing-file.js:
```diff
@@ -0,0 +1,39 @@
+const args = process.argv.slice( 2 );
+if ( args.length < 1 ) {
+	console.error( 'failed to rebuild package: at least one filename must be provided' );
+	process.exit( 1 );
```
The doc suggests

> In most situations, it is not actually necessary to call process.exit() explicitly. The Node.js process will exit on its own if there is no additional work pending in the event loop. The process.exitCode property can be set to tell the process which exit code to use when the process exits gracefully.

(https://nodejs.org/api/process.html#process_process_exitcode).

sirbrillig replied:
I think that if I put this code into a function, then I could replace these with `process.exitCode = 1; return;`, but since this is all top-level code currently, this feels cleaner than `process.exitCode = 1; process.exit();`.

scinos commented on bin/build-package-containing-file.js:
```diff
@@ -0,0 +1,39 @@
+const { spawnSync } = require( 'child_process' );
+
+const args = process.argv.slice( 2 );
+if ( args.length < 1 ) {
```
nit: I'd recommend to use `yargs` or `commander` (already present as dependencies) to do the arguments handling.

scinos commented on bin/build-package-containing-file.js:
```diff
@@ -0,0 +1,39 @@
+	process.exit( 1 );
+}
+const packageJSONFile = packageDirectory + '/package.json';
+const rawData = fs.readFileSync( packageJSONFile );
+const parsedData = JSON.parse( rawData );
```
You can just `const parsedData = require(packageJSONFile)`
```suggestion
const parsedData = require(packageJSONFile)
```

sirbrillig replied:
When I do this, I get a fatal error:

```
Error: Cannot find module 'packages/composite-checkout/package.json'
```

scinos replied:
Ah right, it needs a `./` at the front. Interesting that `readFileSync` is ok with it, TIL.

scinos commented on bin/build-package-containing-file.js:
```diff
@@ -0,0 +1,39 @@
+	console.error( 'failed to rebuild package: missing package name' );
+	process.exit( 1 );
+}
+const buildResult = spawnSync( 'yarn', [ 'lerna run prepare', `--scope="${ packageName }"` ], {
```
No need to use `lerna` here, we can do `yarn workspace ${packageName} prepare `

scinos commented on package.json:
```diff
@@ -75,7 +75,8 @@
 		"build-client-stats": "check-npm-client && NODE_ENV=production CALYPSO_ENV=production EMIT_STATS=true PROGRESS=true yarn run build-client-evergreen",
 		"build-packages": "check-npm-client && npx lerna run prepare --stream",
-		"build-packages:watch": "check-npm-client && chokidar \"./packages/**\" -i \"./packages/*/dist/**\" -i \"**.css\" -i \"**.d.ts\" --debounce 1000 --throttle 10000 -c \"yarn run build-packages\"",
+		"build-packages:watch": "check-npm-client && chokidar \"./packages/**\" -i \"./packages/*/dist/**\" -i \"**.css\" -i \"**.d.ts\" -i \"**.md\" -i \"./packages/*/test/**\" -i \"./packages/*/tests/**\" --debounce 1000 --throttle 10000 -c \"yarn run build-packages:rebuild-one {path}\"",
```
I know this was an existing problem, but I think it is important to point it out:

`build-packages` builds anything that has a `prepare` script. This includes some apps in `./apps`. I'd expect `build-packages:watch` to operate on the same set of packages, but in watch mode. 

However, the proposed implementation will limit it to `./packages` directory

sirreal replied:
A quick, simple win is for apps to stop using `prepare`. I don't think there was justification to use that in the first place and they should use `build`.

Nothing should depend on apps, they shouldn't need to be built for the Calypso client to work.

scinos replied:
(Not for this PR)

Following that logic, there may be `./packages` that are not required for Calypso at all.

I wonder if we could drop `build-packages` as a prerequisite and let webpack handle the compilation/transpilation of the used packages (eg: use https://github.com/TypeStrong/ts-loader for TS packages).

sirreal replied:
That is absolutely another direction to test, at the risk of blurring the line between Calypso client and the packages. We've tried to maintain strong isolation with packages, so the Calypso client is effectively just another consumer of the external packages even though they're in the same repo.

I'd love to see exploration in our project setup and TypeScript tooling in particular 🙂

scinos reviewed (changes requested):
Thanks for improving this.

I added a few suggestions here and there, but nothing critical. The only thing missing for me is dropping `lerna` and use `yarn workspaces`.

alshakero commented on bin/build-package-containing-file.js:
```diff
@@ -0,0 +1,39 @@
+	console.error( 'failed to rebuild package: missing filename' );
+	process.exit( 1 );
+}
+const packageDirectoryMatches = fileName.match( /^packages\/[^/]+/ );
+if ( packageDirectoryMatches.length < 1 ) {
+	console.error( 'failed to rebuild package: missing package directory' );
+	process.exit( 1 );
+}
+const packageDirectory = packageDirectoryMatches[ 0 ];
+if ( ! packageDirectory ) {
+	console.error( 'failed to rebuild package: missing package directory contents' );
+	process.exit( 1 );
+}
```
This would fail on Windows (because Windows uses \ not /).. I would do this:

On top of the script, I'd store the packages absolute dir, 

```js
const packagesFolder = path.resolve( '..', 'packages' );
```

Then below I'd do:

```js
// relative path between packages folder and the affected file
// if you modify data-stores/tsconfig.json, the relative path will be
// data-stores/tsconfig.json 
const affectedFileRelativePath = path.relative( packagesFolder, changedFile );

if(affectedFileRelativePath.beginsWith('..')) {
	console.error( 'Path is outside packages directory' );
    process.exit( 1 );
}

// notice using path.sep, because it's OS agnostic. 
const affectedPackage = relativePath.split( path.sep )[ 0 ];
```

alshakero reviewed (changes requested):
Awesome work. I love this idea, thank you!

Once the Windows issue is fixed, I'll approve right away, I was on Windows for years so I feel for those people :D 

One suggestion that is 100% up to you, is to maybe move the watching logic to the script file you created and get rid of this 
```sh
chokidar \"./packages/**\" -i \"./packages/*/dist/**\" -i \"**.css\" -i \"**.d.ts\" --debounce 1000 --throttle 10000 -c 
```

See how beautiful is this:
```js
const chokidar = require('chokidar');
const watcher = chokidar.watch(packagesFolder, {ignored: /^\./, persistent: true});

watcher
  	.on('change', handleChange)
	.on('add', handleChange)
	.on('unlink', handleChange)
```

sirbrillig committed:
Add yargs to get filename

sirbrillig committed:
Switch lerna run prepare to workspace prepare

sirbrillig committed:
Use path for parsing/building paths

sirbrillig commented:
Thanks, folks! I believe I addressed all the feedback.

> One suggestion that is 100% up to you, is to maybe move the watching logic to the script file you created and get rid of this

I think that's a great idea, but maybe could come in a later PR as it would involve changing the script to be a watcher instead of a rebuilder (which is totally reasonable idea but it feels like a bigger change to me).
